src/index.js

```diff
@@ -335,10 +337,63 @@ class JanusAdapter {
     });
   }
 
+  toggleFreeze() {
+    if(this.frozen) {
+      this.unfreeze();
+    } else {
+      this.freeze();
+    }
+  }
+
+  freeze() {
+    this.frozen = true;
+  }
+
+  unfreeze() {
+    this.frozen = false;
+    this.flushPendingUpdates();
+  }
+
+  flushPendingUpdates() {
+    for (let [networkId, message] of this.frozenUpdates) {
+        if(!message.data.owner || this.occupants[message.data.owner]) { // ignore users who have disconnected since freezing, since their entities will have already been removed by NAF
+          this.onOccupantMessage(null, message.dataType, message.data);
+      }
+    }
+    this.frozenUpdates.clear();
+  }
+
+  storeMessage(message) {
+    const networkId = message.data.networkId;
+    if(!this.frozenUpdates.has(networkId)) {
+      this.frozenUpdates.set(networkId, message);
+    } else {
+      const storedMessage = this.frozenUpdates.get(networkId);
+
+      // Avoid updating components if the entity data received did not come from the current owner.
+      if (message.data.lastOwnerTime < storedMessage.data.lastOwnerTime ||
+            (storedMessage.data.lastOwnerTime === message.data.lastOwnerTime && storedMessage.data.owner > message.data.owner)) {
```

Comment:
Oh man, it feels wrong to leak the ownership algorithm into the adapter. I almost want to move this condition into a shared NAF utility function

Reply:
Yeah, we may as well.

Reply:
Yep this is definitely the ugliest part of the whole thing. Unfortionetly we ened to know this since we are only storing the latest state of each entity while frozen so we have to correctly enforce any ownership transfers. I was trying to keep this update self contained in the adapter since a) I am not sure if this is generally applicable to NAF and b) I don't want to spend the time right now to make NAF pluggable enough to allow implementing this elsewhere. Exposing a utility method might be a reasonable compromise though I think it makes the usage in NAF a bit uglier.

Reply:
Yeah, I think `NAF.utils.isOwner(curLastOwnerTime, curOwner, nextLastOwnerTime, nextOwner)` is the function signature you would need and it's already about half the length of this statement, but it would make refactoring easier.

Reply:
Also worth mentioning that it's likely this algorithm could change too (given the discussion in https://github.com/mozilla/mr-social-client/issues/120), so almost certainly a good idea to centralize it.